src/catalog/database_catalog.cpp

```diff
@@ -1129,14 +1129,17 @@ bool DatabaseCatalog::DeleteIndex(const common::ManagedPointer<transaction::Tran
 
   // Everything succeeded from an MVCC standpoint, so register a deferred action for the GC to delete the index with txn
   // manager. See base function comment.
-  txn->RegisterCommitAction([=](transaction::DeferredActionManager *deferred_action_manager) {
-    deferred_action_manager->RegisterDeferredAction([=]() {
-      deferred_action_manager->RegisterDeferredAction([=]() {
-        delete schema_ptr;
-        delete index_ptr;
+  txn->RegisterCommitAction(
+      [=, garbage_collector{garbage_collector_}](transaction::DeferredActionManager *deferred_action_manager) {
+        if (index_ptr->Type() == storage::index::IndexType::BWTREE)
+          garbage_collector->UnregisterIndexForGC(common::ManagedPointer(index_ptr));
```

Comment:
Perhaps braces and/or a newline to make it explicit that the double-deferral doesn't depend on the index type, though I could just be too sleepy to read carefully right now.